src/adc.rs

```diff
@@ -0,0 +1,359 @@
+        trigger_channel: sampling_timer::Timer2Channel1,
+    ) -> Self {
+        // Generate DMA events when an output compare of the timer hitting zero (timer roll over)
+        // occurs.
+        trigger_channel.listen_dma();
+        trigger_channel.to_output_compare(0);
+
+        // The trigger stream constantly writes to the TX FIFO using a static word (dont-care
+        // contents). Thus, neither the memory or peripheral address ever change. This is run in
+        // circular mode to be completed at every DMA request.
+        let trigger_config = DmaConfig::default()
+            .memory_increment(false)
+            .peripheral_increment(false)
+            .priority(Priority::High)
+            .circular_buffer(true);
+
+        // Construct the trigger stream to write from memory to the peripheral.
+        let mut trigger_transfer: Transfer<_, _, MemoryToPeripheral, _> =
+            Transfer::init(
+                trigger_stream,
+                SPI2::new(),
+                unsafe { &mut SPI_START },
+                None,
+                trigger_config,
+            );
+
+        // The data stream constantly reads from the SPI RX FIFO into a RAM buffer. The peripheral
+        // stalls reads of the SPI RX FIFO until data is available, so the DMA transfer completes
+        // after the requested number of samples have been collected. Note that only ADC1's data
+        // stream is used to trigger a transfer completion interrupt.
+        let data_config = DmaConfig::default()
+            .memory_increment(true)
+            .priority(Priority::VeryHigh)
+            .peripheral_increment(false);
+
+        // A SPI peripheral error interrupt is used to determine if the RX FIFO overflows. This
+        // indicates that samples were dropped due to excessive processing time in the main
+        // application (e.g. a second DMA transfer completes before the first was done with
+        // processing). This is used as a flow control indicator to guarantee that no ADC samples
+        // are lost.
+        let mut spi = spi.disable();
+        spi.listen(hal::spi::Event::Error);
+
+        // The data transfer is always a transfer of data from the peripheral to a RAM buffer.
+        let mut data_transfer: Transfer<_, _, PeripheralToMemory, _> =
+            Transfer::init(
+                data_stream,
+                spi,
+                unsafe { &mut ADC0_BUF0 },
+                None,
+                data_config,
+            );
+
+        data_transfer.start(|spi| {
+            // Allow the SPI FIFOs to operate using only DMA data channels.
+            spi.enable_dma_rx();
+            spi.enable_dma_tx();
+
+            // Enable SPI and start it in infinite transaction mode.
+            spi.inner().cr1.modify(|_, w| w.spe().set_bit());
+            spi.inner().cr1.modify(|_, w| w.cstart().started());
+        });
+
+        trigger_transfer.start(|_| {});
```

Comment:
Because we need to actually start the SPI peripheral there. The trigger transfer start just starts allowing the timer overflow to schedule SPI transfers. If the SPI isn't enabled, it won't do anything.

Reply:
Ah, also  `trigger_transfer` doesn't have ownership of the SPI registers, so we can't do it in that closure.

Reply:
After further discussion, this isn't possible for the DACs because SPI is owned by the transfer. For symmetry, we will follow the API from the DACs and keep it as-is.